rclcpp/include/rclcpp/any_subscription_callback.hpp

```diff
@@ -82,6 +88,38 @@ struct AnySubscriptionCallback
   {
     shared_ptr_with_info_callback = callback;
   }
+
+  template<typename CallbackT,
+  typename std::enable_if<
+    function_traits<CallbackT>::arity == 1
+  >::type * = nullptr,
+  typename std::enable_if<
+    std::is_same<
+      typename function_traits<CallbackT>::template argument_type<0>,
+      typename std::shared_ptr<const MessageT>
+    >::value
+  >::type * = nullptr
+  >
+  void set(CallbackT callback)
+  {
+    const_shared_ptr_callback = callback;
+  }
+
+  template<typename CallbackT,
+  typename std::enable_if<
+    function_traits<CallbackT>::arity == 2
+  >::type * = nullptr,
+  typename std::enable_if<
+    std::is_same<
+      typename function_traits<CallbackT>::template argument_type<0>,
+      typename std::shared_ptr<const MessageT>
+    >::value
+  >::type * = nullptr
+  >
+  void set(CallbackT callback)
+  {
```

Comment:
The new test passes: http://ci.ros2.org/job/ros2_batch_ci_windows/460/testReport/%28root%29/test_subscription__rmw_connext_cpp/subscription_shared_ptr_const/

I'm still not convinced that we need a static assert, maybe we should chat about it offline?